### ISSUE TITLE
Failing on startup on pending migrations is probably not a good idea

### ISSUE THREAD

**Issue by ProGM** [neo4jrb/neo4j#1295]
Fail on startup when there are pending migrations in rails is not a great idea.
That blocks any kind of rake task, except `neo4j:migrate`. Moreover, the code that excludes that task is very unclear and buggy (not reusable in other rails tasks, for examples).

Probably it would be better to move to the AR strategy: startup correctly, then fail on every request to the server. What do you think @subvertallchris @cheerfulstoic ?

### Runtime information:

Neo4j database version:
`neo4j` gem version: 8.0.0
`neo4j-core` gem version: 7.0.0

**Comment by cheerfulstoic** [comment]
That makes sense to me.  By "fail on every request to the server" you mean Neo4j, right?

**Comment by ProGM** [comment]
wouldn't be failing on every database request pretty expensive?

Uhm... No, I mean, failing exactly like AR does: using a rack middleware that checks for pending migrations.

Like now, there will be a method to be called to check for pending migrations in non-rails environments ;P

**Comment by cheerfulstoic** [comment]
I don't think it would necessarily be expensive.  We could have a cached variable that we check to see if migrations have been run.

That said, I say do whatever ActiveRecord is doing ;)

**Comment by subvertallchris** [comment]
I'm with Brian: do whatever ActiveRecord is doing. I do regret that this would prevent someone who is not using Rack from taking advantage of this protection, though.

**Comment by ProGM** [comment]
@subvertallchris 
who's not using rack could simply call `Neo4j::Migrations.check_for_pending_migrations!` in his application ;)

**Comment by subvertallchris** [comment]
👍

**Comment by cheerfulstoic** [comment]
I have some thoughts on how to refactor `SessionManager` so that it can establish sessions for non-Rails apps.  I wonder if we could add a bit to check the migrations in there too so that non-Rails apps only need to call one thing

**Comment by ProGM** [comment]
👍  Reopening to remember about this
